src/form/middleware.ts

```diff
@@ -0,0 +1,195 @@
+		(callback: (values: S) => void, defaults: S): void;
+	};
+	onValue(callback: (values: Partial<S>) => void): void;
+	disabled(disabled?: boolean): boolean;
+	valid(): boolean;
+	reset(): void;
+	field<K extends keyof S>(name: K, required?: boolean): Field<S, K>;
+}
+
+export interface Field<S, K extends keyof S> {
+	value: {
+		(newValue: S[K]): S[K];
+		(): S[K] | undefined;
+	};
+	valid(valid?: boolean, message?: string): Validity;
+	required(required?: boolean): boolean;
+	disabled(required?: boolean): boolean;
+}
+
+function compareValidity(a: Validity, b: Validity) {
+	if (typeof a === 'boolean' && typeof b === 'boolean') {
+		return true;
+	}
+	if (typeof a === 'boolean' || typeof b === 'boolean') {
+		return false;
+	}
+	if (b.valid === undefined) {
+		return false;
+	}
+	return a.message === b.message;
+}
+
+interface State<S> {
+	required: Required;
+	values: Partial<S>;
+	valid: Valid;
+	disabled: Disabled;
+	formDisabled: boolean;
+}
+
+export const createFormMiddleware = <S extends Record<string, any> = any>() => {
+	const icache = createICacheMiddleware<State<S>>();
+	const factory = create({ icache, cache });
+
+	const formMiddleware = factory(function Form({
+		middleware: { icache, cache }
+	}): FormMiddleware<S> {
+		return {
+			value(values?: any): any {
+				const currentValues = icache.getOrSet('values', {});
+				if (values !== undefined) {
+					icache.set('values', values);
+					const newValues = { ...currentValues, ...values };
+					const callbacks = cache.get<Callbacks<S>>('callbacks');
+					callbacks && callbacks.onValue && callbacks.onValue(newValues);
+					return newValues;
+				}
+				return currentValues;
+			},
+			submit<T = void>(callback: (values: any) => T, defaults?: S) {
+				if (!this.valid()) {
+					return;
+				}
+				if (defaults) {
+					return callback({ ...defaults, ...this.value() } as S);
+				}
+				return callback(this.value());
+			},
+			onValue<T = void>(callback: (values: Partial<S>) => T) {
+				cache.set('callbacks', { onValue: callback });
+			},
+			disabled(disabled?: boolean) {
+				if (disabled !== undefined) {
+					icache.set('formDisabled', disabled);
+					return disabled;
+				}
+				return icache.getOrSet('formDisabled', false);
+			},
+			valid() {
+				const values = icache.getOrSet('valid', {});
+				const requiredValues = icache.getOrSet('required', {});
+				return Object.keys(values).every((key) => {
+					const valid = values[key];
+					const value = typeof valid === 'boolean' ? valid : valid.valid;
+					return (value === undefined && !requiredValues[key]) || Boolean(value);
+				});
+			},
+			reset() {
+				icache.set('values', {});
+				let valid: Valid = {};
+				const values = icache.getOrSet('valid', {});
+				Object.keys(values).map((key) => {
+					valid[key] = { valid: undefined, message: '' };
+				});
+				icache.set('valid', valid);
+				icache.set('required', {});
+			},
+			field<K extends keyof S>(name: K, required = false): Field<S, K> {
```

Comment:
can name / `k` be changed to extend or use typeof string here to avoid the use of `as string` elsewhere?